cpp/ql/src/semmle/code/cpp/PrintAST.qll

```diff
@@ -114,9 +114,32 @@ class PrintASTNode extends TPrintASTNode {
 
   /**
    * Gets the child node at index `childIndex`. Child indices must be unique,
-   * but need not be contiguous (but see `getChildByRank`).
+   * but need not be contiguous.
    */
-  abstract PrintASTNode getChild(int childIndex);
+  abstract PrintASTNode getChildInternal(int childIndex);
+
+  /**
+   * Gets the child node at index `childIndex`.
+   * Adds edges to fully converted expressions, that are not part of the
+   * regular parent/child relation traversal.
+   */
+  PrintASTNode getChild(int childIndex) {
+    result = getChildInternal(childIndex)
+    or
+    // We first compute the first available child index that is not used by
+    // `getChildInternal`, then we synthesize the child for fully converted
+    // expressions at `nextAvailableIndex` plus the childIndex of the non-converted
+    // expression. This ensures that both disjuncts are disjoint.
+    exists(int nonConvertedIndex, int nextAvailableIndex, Expr expr |
+      nextAvailableIndex = max(int idx | exists(this.getChildInternal(idx))) + 1 and
+      childIndex - nextAvailableIndex = nonConvertedIndex and
+      expr = getChild(nonConvertedIndex).(ASTNode).getAST()
```

Comment:
Is it necessary to call `getChild` here rather than `getChildInternal`? I don't see why this predicate needs to be recursive.